strawberry/schema/execute_context.py

```diff
@@ -0,0 +1,127 @@
+
+from promise import Promise, is_thenable, promise_for_dict
+
+from graphql import (
+    ExecutionContext,
+    GraphQLError,
+    GraphQLField,
+    GraphQLFieldResolver,
+    GraphQLObjectType,
+    GraphQLOutputType,
+    GraphQLResolveInfo,
+)
+from graphql.execution.values import get_argument_values
+from graphql.language import FieldNode
+from graphql.pyutils import AwaitableOrValue, Path, Undefined
+
+
+class ExecutionContextWithPromise(ExecutionContext):
+    def build_response(self, data):
+        if is_thenable(data):
+            original_build_response = super().build_response
+
+            def on_rejected(error):
+                self.errors.append(error)
+                return None
+
+            def on_resolve(data):
+                return original_build_response(data)
+
+            promise = data.catch(on_rejected).then(on_resolve)
+            return promise
+        return super().build_response(data)
+
+    def resolve_field_value_or_error(
+        self,
+        field_def: GraphQLField,
+        field_nodes: List[FieldNode],
+        resolve_fn: GraphQLFieldResolver,
+        source: Any,
+        info: GraphQLResolveInfo,
+    ) -> Union[Exception, Any]:
+        """Resolve field to a value or an error.
+
+        Isolates the "ReturnOrAbrupt" behavior to not de-opt the resolve_field()
+        method. Returns the result of resolveFn or the abrupt-return Error object.
+
+        For internal use only.
+        """
+        try:
+            # Build a dictionary of arguments from the field.arguments AST, using the
+            # variables scope to fulfill any variable references.
+            args = get_argument_values(field_def, field_nodes[0], self.variable_values)
+
+            # Note that contrary to the JavaScript implementation, we pass the context
+            # value as part of the resolve info.
+            result = resolve_fn(source, info, **args)
+            if is_thenable(result):
+                return result
+
+            return result
+        except GraphQLError as error:
+            return error
+        except Exception as error:
+            return GraphQLError(str(error), original_error=error)
+
+    def complete_value_catching_error(
+        self,
+        return_type: GraphQLOutputType,
+        field_nodes: List[FieldNode],
+        info: GraphQLResolveInfo,
+        path: Path,
+        result: Any,
+    ) -> AwaitableOrValue[Any]:
+        """Complete a value while catching an error.
+        This is a small wrapper around completeValue which detects and logs errors in
+        the execution context.
+        """
+        completed: AwaitableOrValue[Any]
+        try:
+            if is_thenable(result):
+
+                def handle_error(error):
+                    self.handle_field_error(error, field_nodes, path, return_type)
+
+                completed = Promise.resolve(result).then(
+                    lambda resolved: self.complete_value(
```

Comment:
Just for consistency, i would define both args as lambdas or both as 'normal' functions

Reply:
Yep agreed 👍